tools/SwiftSyntax/SyntaxData.swift

```diff
@@ -198,3 +257,62 @@ final class SyntaxData: Equatable {
     return lhs === rhs
   }
 }
+
+/// An absolute position in a source file as text - the absolute byteOffset from
+/// the start, line, and column.
+public class AbsolutePosition {
+  public private(set) var byteOffset: Int = 0
+  public private(set) var line: Int = 1
+  public private(set) var column: Int = 1
+
+  internal func copy() -> AbsolutePosition {
+    let result = AbsolutePosition()
+    result.byteOffset = byteOffset
+    result.line = line
+    result.column = column
+    return result
+  }
+
+  /// Add some number of columns to the position.
+  internal func add(columns: Int) {
+    column += columns
+    byteOffset += columns
+  }
+
+  /// Add some number of newlines to the position, resetting the column.
+  /// Size is byte size of newline char.
+  /// '\n' and '\r' are 1, '\r\n' is 2.
+  internal func add(lines: Int, size: Int) {
+    line += lines
+    column = 1
+    byteOffset += lines * size
+  }
+
+  /// Use some text as a reference for adding to the absolute position,
+  /// taking note of newlines, etc.
+  internal func add(text: String) {
+    guard let chars = text.cString(using: .utf8) else {
```

Comment:
IIRC This needs to be `free`d. You might be able to avoid C strings altogether, and this optional, by using `text.utf8`.